Source/LogMessage.swift

```diff
@@ -25,10 +25,15 @@
 import Foundation
 
 /// A LogMessage is a detailed log entry with a name and a dictionary of associated attributes.
-public protocol LogMessage {
+public protocol LogMessage: CustomStringConvertible {
     /// Name of this message.
     var name: String { get }
 
     /// Attributes associated with this message.
     var attributes: [String: Any] { get }
 }
+extension LogMessage  {
+    public var description: String {
+        return "\(name): \(attributes)"
+    }
+}
```

Comment:
While this is certainly clever, it is very limiting and defeats the main reason the `LogMessage` protocol was added in the first place. The `LogMessage` protocol allows you to make more robust types and to split out the information so a `LogWriter` can do something different with it. For example, they could send `LogMessage` types to NewRelic, or switch on certain attributes to report errors in a different way. There are really no limits to what you can do with the attributes as long as they are maintained all the way through to the `LogWriter` instances.

This change converts the `name`-`attributes` pair into a `String` which is then passed through Willow. Therefore, by the time the `LogWriter` instances get the `LogMessage`, it is now a `String` and it can't key off the `attributes` to do custom things.

Again, this is a clever solution that certainly removes ambiguity, but also removes a core feature of the Willow 4 release.

Reply:
As Christian says, the important part of the change was to be able to route additional (non-string) data to writers. Typically this would be context info, etc. In the case of a Console or other text writer, yes, collapsing to a single String can work.

However, when the writer wants to keep the attributes as a separate data entity that can be used for filtering, faceting, and aggregating, reducing to a string makes that impossible.

I do think there could be benefit to having `Message` conform to `CustomStringConvertible` for debugging purposes, assuming that didn't create more ambiguity with String types.

Reply:
Actually that's not what's happening. 

I think you're focusing on the fact that CustomStringConvertible *can necessarily produce a String*, but that doesn't mean that *it is a String*.  

Also, ll. 36-38 above is just a default implementation that can be replaced with whatever you want. 

You can indeed have your cake and eat it to with this, since the LogMessage itself (and all its glory) is a CustomStringConvertible. The LogWriter can still introspect into its attributes by simply doing `if let logMessage = message as? LogMessage` ... see the updated test case I just pushed.

Reply:
Alright this was closed, I guess I have to open another PR.